### PR TITLE
Add `Polynormer` model and example [1/n]

### PR DESCRIPTION
**Pull request by xnuohz**
### Issue
- From #9751 

### Usage
```python examples/polynormer.py --post_bn```

### Highlight
image
image

First round of benchmarks run on NVIDIA internal computers

### CODE REVIEW

**Review by puririshi98 — CHANGES_REQUESTED** [review]
please share a log of running the example

**Review by puririshi98 — CHANGES_REQUESTED** [review]
please put the Model from example into torch_geometric.nn.models
then also put a unit test under test/nn/models. 
then please integrate the model into examples/ogbn_train.py.
do this with "--gnn-choice" arg parse option, with choices ["sage, gat, polynormer_graph_transformer"]

please share the log of all 3 and then change the default to the best
(i will have other PRs for GraphTrasnformers follow this design pattern so its unified)

**Comment by xnuohz** [comment]
| file | model  | valid acc | test acc | avg training time | avg inference time |
| -- | -- | -- | -- | -- | -- |
| ogbn_train | Polynormer | 72.81% | 71.18% | 29.1350 | 4.3969 |
| ogbn_train_cugraph | Polynormer | 72.76% | 70.95% | 32.5175 | 23.7977 |
| ogbn_train_cugraph_multigpu | Polynormer | 71.35% | 70.24% | 32.2606 | 23.8230 |

[ogbn_train_cugraph_multi_gpu_polynormer.log](https://github.com/user-attachments/files/18824746/ogbn_train_cugraph_multi_gpu_polynormer.log)
[ogbn_train_cugraph_polynormer.log](https://github.com/user-attachments/files/18824747/ogbn_train_cugraph_polynormer.log)
[ogbn_train_polynormer.log](https://github.com/user-attachments/files/18824748/ogbn_train_polynormer.log)

File: torch_geometric/nn/models/polynormer.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        dropout: float = 0.5,
+        global_dropout: float = 0.5,
+        heads: int = 1,
+        beta: float = 0.9,
+        pre_ln: bool = False,
+        post_bn: bool = True,
+        local_attn: bool = False,
+    ):
+        super().__init__()
+        self._global = True
```
Comment (xnuohz) [inline torch_geometric/nn/models/polynormer.py:56]: @puririshi98  the paper's training method involves warming up local attention for a few epochs and switching to global attention. however, to be consistent with the training methods of several other models in 3 files, I removed the warmup process and still got good results. so I think this can allow the model's code further simplified, and only retain global attention. what do you think?

File: torch_geometric/nn/attention/polynormer.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            q = F.sigmoid(self.q(x)).view(N, self.head_channels, self.heads)
+        v = self.v(x).view(N, self.head_channels, self.heads)
+
+        # numerator
+        kv = torch.einsum('ndh, nmh -> dmh', k, v)
+        num = torch.einsum('ndh, dmh -> nmh', q, kv)
+
+        # denominator
+        k_sum = torch.einsum('ndh -> dh', k)
+        den = torch.einsum('ndh, dh -> nh', q, k_sum).unsqueeze(1)
```
Comment (xnuohz) [inline torch_geometric/nn/attention/polynormer.py:68]: also not attend across batch, need to fix

File: torch_geometric/nn/models/polynormer.py (hunk 1 above)
Comment (puririshi98) [reply torch_geometric/nn/models/polynormer.py:60]: i think its best to go by the original paper and to just add the extra code under if args.gnn_name == "polynormer":
do special stuff

**Comment by xnuohz** [comment]
1. Polynormer also supports batch
2. update ogbn_train.py and follow the paper's training style
3. add an error message for the cugraph example before cugraph support disjoint
4. my resource is limited, so batching graph transformer is hard to train and get a result

File: torch_geometric/nn/models/polynormer.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        for _ in range(global_layers):
+            self.global_attn.append(
+                PolynormerAttention(
+                    hidden_channels,
+                    heads,
+                    hidden_channels,
+                    beta,
+                    global_dropout,
+                ))
+        self.pred_local = torch.nn.Linear(inner_channels, out_channels)
```
Comment (puririshi98) [inline torch_geometric/nn/models/polynormer.py:129]: need to put this behind        if/else not self._global: (same with in the reset params) since each is only used in a certain case

**Review by puririshi98 — CHANGES_REQUESTED** [review]

File: examples/ogbn_train.py [hunk 4]
```diff
@@ -34,17 +34,22 @@
...
-    "--model",
-    type=str.lower,
-    default='SGFormer',
-    choices=['sage', 'gat', 'sgformer'],
-    help="Model used for training",
+    '--gnn_choice',
+    type=str,
+    default='graph_transformer_sgformer',
+    choices=[
+        'gat', 'sage', 'graph_transformer_sgformer',
```
Comment (puririshi98) [inline examples/ogbn_train.py:41]: please switch this to "SGFormer" and "Polynormer" and then below, keep a hardcoded list of which choices are graph transformer, then you can do args.gnn_choice = "graph_transformer_"+ args.gnn_choice.
please also mirror all of this to the cugraph examples including the appropriate changes to the forward pass (imagine sampling worked, since in the future it will and we want this to be the base reference).

**Review by puririshi98 — CHANGES_REQUESTED** [review]

File: torch_geometric/nn/models/polynormer.py (hunk 3 above)
Comment (xnuohz) [reply torch_geometric/nn/models/polynormer.py:129]: wdym? in a complete training, both pred_global and pred_local will be used, so should init both. warmup in first 50 epochs -> pred_local, switch to last 50 epochs -> pred_global

Comment (puririshi98) [reply torch_geometric/nn/models/polynormer.py:129]: oh i see i misunderstood had not fully read the code yet, i will do a deeper review soon

**Comment by puririshi98** [comment]
@Kh4L any reviews welcome as well

File: torch_geometric/nn/attention/polynormer.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        # numerator
+        kv = torch.einsum('bndh, bnmh -> bdmh', k, v)
+        num = torch.einsum('bndh, bdmh -> bnmh', q, kv)
+
+        # denominator
+        k_sum = torch.einsum('bndh -> bdh', k)
+        den = torch.einsum('bndh, bdh -> bnh', q, k_sum).unsqueeze(2)
+
+        # linear global attention based on kernel trick
+        x = (num / den).reshape(B, N, -1)
```
Comment (Kh4L) [inline torch_geometric/nn/attention/polynormer.py:85]: we might want to add some eps in den for numerical stability

File: torch_geometric/nn/attention/polynormer.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+    def forward(self, x: Tensor, mask: Optional[Tensor] = None) -> Tensor:
+        r"""Forward pass.
+
+        Args:
+            x (torch.Tensor): Node feature tensor
+                :math:`\mathbf{X} \in \mathbb{R}^{B \times N \times F}`, with
+                batch-size :math:`B`, (maximum) number of nodes :math:`N` for
+                each graph, and feature dimension :math:`F`.
+            mask (torch.Tensor, optional): Mask matrix
```
Comment (Kh4L) [inline torch_geometric/nn/attention/polynormer.py:63]: unused parameter?

File: torch_geometric/nn/attention/polynormer.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        """
+        B, N, *_ = x.shape
+        h = self.h_lins(x)
+        k = F.sigmoid(self.k(x)).view(B, N, self.head_channels, self.heads)
+        if self.qk_shared:
+            q = k
+        else:
+            q = F.sigmoid(self.q(x)).view(B, N, self.head_channels, self.heads)
+        v = self.v(x).view(B, N, self.head_channels, self.heads)
+
```
Comment (Kh4L) [inline torch_geometric/nn/attention/polynormer.py:75]: I believe it should be:
```
        if mask is not None:
            mask = mask[:, :, None, None]
            v.masked_fill_(~mask, 0.)
```

**Review by Kh4L — CHANGES_REQUESTED** [review]
Nice contribution! Left a few comments, main concern is the unused mask param for batching

**Comment by puririshi98** [comment]
> the 3-layer polynormer is not as good as gcn, while the original paper uses 7 layers [ogbn_polynormer.log](https://github.com/user-attachments/files/19108763/ogbn_polynormer.log)

@xnuohz  since this example framewokr is not just for polynormer lets leave default at 3, and add an if model==polynormer and num_layers!=7: print("The original polynormer paper recommends 7 layers, you have chosen", num_layers, " which may effect results. See <insert paper link> for details")

File: examples/multi_gpu/ogbn_train_cugraph.py [hunk 8]
```diff
@@ -80,11 +82,11 @@ def arg_parse():
         help='Whether or not to add self loop',
     )
     parser.add_argument(
-        "--model",
+        "--gnn_choice",
```
[the remark this answers is not in the document]
Comment (xnuohz) [reply examples/multi_gpu/ogbn_train_cugraph.py:84]: oh, they are consistent with the name `gnn_choice`, do you think I need to change it back as `model`?

**Comment by xnuohz** [comment]
@puririshi98 @Kh4L ready for review, thanks:)

File: examples/multi_gpu/ogbn_train_cugraph.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Kh4L) [reply examples/multi_gpu/ogbn_train_cugraph.py:84]: Both work fine as long as it is consistent!

**Review by Kh4L — APPROVED** [review]

**Comment by puririshi98** [comment]
need to test and then ill ask matthias for review to avoid similar mistakes to the last PR

**Comment by xnuohz** [comment]
[ogbn_polynormer.log](https://github.com/user-attachments/files/19321163/ogbn_polynormer.log)

**Review by puririshi98 — CHANGES_REQUESTED** [review]
actually i think we should wait until @alexbarghi-nv has disjoint sampling ready which is not too far off for the cugraph functionality, can you split this into two PRs @xnuohz

**Comment by xnuohz** [comment]
@puririshi98 i see. so i need to move cugraph example related updates out of this PR and make a new one, right?

**Comment by xnuohz** [comment]
@puririshi98 ready for review

**Review by puririshi98 — CHANGES_REQUESTED** [review]
lgtm, lets see if @akihironitta or @rusty1s has any concerns

File: examples/ogbn_train.py [hunk 9]
```diff
@@ -97,36 +105,23 @@
...
+
+def get_loader(input_nodes: dict[str, Tensor]) -> NeighborLoader:
+    return NeighborLoader(
+        data,
+        input_nodes=input_nodes,
+        num_neighbors=[args.fan_out] * num_layers,
+        batch_size=batch_size,
+        shuffle=True,
+        num_workers=num_workers,
+        persistent_workers=True,
```
Comment (akihironitta) [inline examples/ogbn_train.py:117]:
```suggestion
        persistent_workers=num_workers > 0,
```

File: torch_geometric/nn/models/polynormer.py [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+            x = F.relu(x)
+            x = F.dropout(x, p=self.dropout, training=self.training)
+            x = (1 - self.beta) * self.lns[i](h * x) + self.beta * x
+            x_local = x_local + x
+
+        # equivariant global attention
+        if self._global:
+            batch, indices = batch.sort()
+            rev_perm = torch.empty_like(indices)
+            rev_perm[indices] = torch.arange(len(indices)).to(indices.device)
```
Comment (akihironitta) [inline torch_geometric/nn/models/polynormer.py:191]:
```suggestion
            rev_perm[indices] = torch.arange(len(indices), device=indices.device)
```

**Review by akihironitta — COMMENTED** [review]
Just skimmed through and haven't had a chance to review the model architecture, but I'll leave it to @puririshi98 on that.

File: torch_geometric/nn/attention/polynormer.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+            (default: :obj:`64.`)
+        beta (float, optional): Polynormer beta initialization.
+            (default: :obj:`0.9`)
+        qkv_bias (bool, optional): If specified, add bias to query, key
+            and value in the self attention. (default: :obj:`False`)
+        qk_shared (bool optional): Whether weight of query and key are shared.
+            (default: :obj:`True`)
+        dropout (float, optional): Dropout probability of the final
+            attention output. (default: :obj:`0.0`)
+
```
Comment (rusty1s) [inline torch_geometric/nn/attention/polynormer.py:26]:
```suggestion
```

File: torch_geometric/nn/attention/polynormer.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+                :math:`\mathbf{X} \in \mathbb{R}^{B \times N \times F}`, with
+                batch-size :math:`B`, (maximum) number of nodes :math:`N` for
+                each graph, and feature dimension :math:`F`.
+            mask (torch.Tensor, optional): Mask matrix
+                :math:`\mathbf{M} \in {\{ 0, 1 \}}^{B \times N}` indicating
+                the valid nodes for each graph. (default: :obj:`None`)
+        """
+        B, N, *_ = x.shape
+        h = self.h_lins(x)
+        k = F.sigmoid(self.k(x)).view(B, N, self.head_channels, self.heads)
```
Comment (rusty1s) [inline torch_geometric/nn/attention/polynormer.py:69]: no need to use `F.sigmoid`, just use `tensor.sigmoid()`

**Review by rusty1s — APPROVED** [review]
would be great to increase code coverage.

**Comment by puririshi98** [comment]
@xnuohz can you see if you can fix the CI issue it might be that the explainer test runs somethjing based on models available in pyg and you have added a new model

**Comment by xnuohz** [comment]
@puririshi98 waiting #10310 to fix and merge
also checkout [this pr](https://github.com/pyg-team/pytorch_geometric/pull/10283), same ci error as mine

**Comment by puririshi98** [comment]
> @puririshi98 waiting #10310 to fix and merge also checkout [this pr](https://github.com/pyg-team/pytorch_geometric/pull/10283), same ci error as mine

okay perfect analysis, this isnt on you. i will merge this once those go in and CI becomes green

**Comment by xnuohz** [comment]
hi @puririshi98 CI becomes green, can you merge?

**Review by puririshi98 — CHANGES_REQUESTED** [review]
lgtm, just wondering why there are changes to appnp, what does this have to do with polynormer?

**Comment by xnuohz** [comment]
CI may fail when torch.allclose uses default parameters, changing these default parameters allows CI to pass

**Review by puririshi98 — APPROVED** [review]
